pkg/runner/test_runner_route_test.go

```diff
@@ -0,0 +1,403 @@
+	"github.com/facebookincubator/contest/pkg/test"
+	"github.com/facebookincubator/contest/plugins/storage/memory"
+	"github.com/facebookincubator/contest/plugins/teststeps/example"
+
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+var testSteps = map[string]test.TestStepFactory{
+	example.Name: example.New,
+}
+
+var testStepsEvents = map[string][]event.Name{
+	example.Name: example.Events,
+}
+
+type TestRunnerSuite struct {
+	suite.Suite
+	routingChannels routingCh
+	router          *router
+
+	// the following channels are the same channels wrapped within `routingChannels`,
+	// but used in the opposite direction compared to the routing block. When the
+	// routing block reads results, the corresponding channel of the test suite has
+	// write direction, when the routing block writes results, the corresponding channel
+	// of the test suite has read direction
+	routeInCh  chan<- *target.Target
+	routeOutCh <-chan *target.Target
+
+	stepInCh  <-chan *target.Target
+	stepOutCh chan<- *target.Target
+	stepErrCh chan<- cerrors.TargetError
+
+	targetErrCh <-chan cerrors.TargetError
+}
+
+func (suite *TestRunnerSuite) SetupTest() {
+
+	log := logging.GetLogger("TestRunnerSuite")
+
+	pluginRegistry := pluginregistry.NewPluginRegistry()
+	// Setup the PluginRegistry by registering TestSteps
+	for name, tsfactory := range testSteps {
+		if _, ok := testStepsEvents[name]; !ok {
+			suite.T().Errorf("test step %s does not define any associated event", name)
+		}
+		err := pluginRegistry.RegisterTestStep(name, tsfactory, testStepsEvents[name])
+		require.NoError(suite.T(), err)
+	}
+	ts1, err := pluginRegistry.NewTestStep("Example")
+	require.NoError(suite.T(), err)
+
+	params := make(test.TestStepParameters)
+	bundle := test.TestStepBundle{TestStep: ts1, TestStepLabel: "FirstStage", Parameters: params}
+
+	routeInCh := make(chan *target.Target)
+	routeOutCh := make(chan *target.Target)
+	stepInCh := make(chan *target.Target)
+	stepOutCh := make(chan *target.Target)
+	stepErrCh := make(chan cerrors.TargetError)
+	targetErrCh := make(chan cerrors.TargetError)
+
+	suite.routeInCh = routeInCh
+	suite.routeOutCh = routeOutCh
+	suite.stepOutCh = stepOutCh
+	suite.stepInCh = stepInCh
+	suite.stepErrCh = stepErrCh
+	suite.targetErrCh = targetErrCh
+
+	suite.routingChannels = routingCh{
+		routeIn:   routeInCh,
+		routeOut:  routeOutCh,
+		stepIn:    stepInCh,
+		stepErr:   stepErrCh,
+		stepOut:   stepOutCh,
+		targetErr: targetErrCh,
+	}
+
+	s, err := memory.New()
+	require.NoError(suite.T(), err)
+	storage.SetStorage(s)
+
+	header := testevent.Header{
+		JobID:         1,
+		RunID:         1,
+		TestName:      "TestRunnerSuite",
+		TestStepLabel: "TestRunnerSuite",
+	}
+	ev := storage.NewTestEventEmitterFetcher(header)
+
+	timeouts := TestRunnerTimeouts{
+		StepInjectTimeout:   30 * time.Second,
+		MessageTimeout:      5 * time.Second,
+		ShutdownTimeout:     1 * time.Second,
+		StepShutdownTimeout: 1 * time.Second,
```

Comment:
Couldn't so short timeout cause a false-positive problem in tests? (because of swapping on the testing node, for example). May be we should make all this timeouts infinite (and also remove `time.After` below)?

Reply:
I consider some of these timeouts important, and the initial idea was not to overcomplicate the design by allowing per step/test timeouts. I'll give some background on those timeouts:
* MessageTimeout: is the timeout for communication within ConTest framwork. At any point in time, it should be guaranteed that ConTest is listening for a message on the receiving side. If a sender hangs, it's the indication of a bug. If this timeout fires, the framework panics.
* ShutdownTimeout: this is maximum time allowed for the pipeline to terminate after a cancellation signal. Also here, it's expected that all Steps terminate after a cancellation signal is delivered. If that doesn't happen, we want to flag the Steps as non-compliant
* StepShutdownTimeout: this is a more granular timeout when waiting for test steps when all targets have terminated. Also here, we expect the steps to terminate within a reasonable timeout.


EDIT:
I see you are specifically referring to the usage in test environment. I think for the reason above, these timeouts should be there to clearly fail the test. Possibly higher to prevent as much as possible false-positives, but I don't think they should be infinite. I'll increase by a factor 5.